experiments/distributedshell/src/main/scala/org/apache/gearpump/experiments/cluster/appmaster/AbstractAppMaster.scala

```diff
@@ -17,76 +17,96 @@
-import org.apache.gearpump.cluster.MasterToAppMaster.{AppMasterRegistered, ResourceAllocated}
-import org.apache.gearpump.cluster.appmaster.ExecutorSystem
-import org.apache.gearpump.cluster.scheduler.Resource
+import com.typesafe.config.{Config, ConfigFactory}
+import org.apache.gearpump.cluster.AppMasterToMaster.GetAllWorkers
+import org.apache.gearpump.cluster.ClientToMaster.ShutdownApplication
+import org.apache.gearpump.cluster.MasterToAppMaster.WorkerList
+import org.apache.gearpump.cluster.appmaster.ExecutorSystemScheduler.{StartExecutorSystemTimeout, StartExecutorSystems, ExecutorSystemStarted, ExecutorSystemJvmConfig}
 import org.apache.gearpump.cluster._
-import org.apache.gearpump.experiments.cluster.AppMasterToExecutor.LaunchTask
-import org.apache.gearpump.experiments.cluster.ExecutorToAppMaster.RegisterExecutor
-import org.apache.gearpump.experiments.cluster.executor.{TaskLaunchData, DefaultExecutor}
-import org.apache.gearpump.experiments.cluster.task.TaskContext
-import org.apache.gearpump.util.ActorSystemBooter.BindLifeCycle
-import org.apache.gearpump.util.{ActorSystemBooter, ActorUtil, Util, LogUtil}
+import org.apache.gearpump.cluster.scheduler.{Relaxation, Resource, ResourceRequest}
+import org.apache.gearpump.experiments.cluster.AppMasterToExecutor.MsgToExecutor
+import org.apache.gearpump.experiments.cluster.ExecutorToAppMaster.{ResponsesFromExecutor, RegisterExecutor}
+import org.apache.gearpump.experiments.cluster.util.ResponseBuilder
+import org.apache.gearpump.util.{Constants, Util, LogUtil}
 import org.slf4j.Logger
 
-import scala.collection.mutable
-import scala.concurrent.duration.{Duration, FiniteDuration}
+import scala.concurrent.Future
+import akka.pattern.{ask, pipe}
 
 abstract class AbstractAppMaster(appContext : AppMasterContext, app : Application) extends ApplicationMaster {
-  import context.dispatcher
   import appContext._
+  import context.dispatcher
+  implicit val timeout = Constants.FUTURE_TIMEOUT
 
   protected val userConfig = app.userConfig
   protected val systemConfig = context.system.settings.config
   protected var master: ActorRef = appContext.masterProxy
   protected var currentExecutorId = 0
-  protected val executorClass: Class[_ <: DefaultExecutor]
-  protected val defaultMsgHandler = masterMsgHandler orElse workerMsgHandler
-
+  protected val defaultMsgHandler = masterMsgHandler orElse workerMsgHandler orElse clientMsgHandler
   private val LOG: Logger = LogUtil.getLogger(getClass, app = appId)
 
+  protected def executorClass: Class[_ <: Actor]
+
   def onStart(): Unit
 
   override def receive: Receive = null
 
   override def preStart(): Unit = {
+    context.become(defaultMsgHandler)
     onStart()
   }
 
   def masterMsgHandler: Receive = {
-    case executorSystem @ ExecutorSystem(_, address, _, executorResource, worker) =>
-      val executorContext = ExecutorContext(currentExecutorId, worker.workerId, appId, self, resource)
+    case ExecutorSystemStarted(executorSystem) =>
```

Comment:
This class is short, it may be eaiser to read by combing all handler to a single one.